src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,391 @@
+                                          uint8_t                       aSeriesId,
+                                          const LinkMetricsTypeIdFlags *aTypeIdFlags,
+                                          uint8_t                       aTypeIdFlagsCount)
+{
+    otError                    error = OT_ERROR_NONE;
+    LinkMetricsQueryOptionsTlv linkMetricsQueryOptionsTlv;
+    uint8_t                    length = 0;
+    static const uint8_t       tlvs[] = {Mle::Tlv::kLinkMetricsReport};
+    uint8_t buf[sizeof(Tlv) * 5 + sizeof(uint8_t) + sizeof(LinkMetricsTypeIdFlags) * kMaxTypeIdFlags];
+    Tlv *   tlv = reinterpret_cast<Tlv *>(buf);
+    Tlv     subTlv;
+
+    // Link Metrics Query TLV
+    tlv->SetType(Mle::Tlv::kLinkMetricsQuery);
+    length += sizeof(Tlv);
+
+    // Link Metrics Query ID sub-TLV
+    subTlv.SetType(kLinkMetricsQueryId);
+    subTlv.SetLength(sizeof(uint8_t));
+    memcpy(buf + length, &subTlv, sizeof(subTlv));
+    length += sizeof(subTlv);
+    memcpy(buf + length, &aSeriesId, sizeof(aSeriesId));
+    length += sizeof(aSeriesId);
+
+    // Link Metrics Query Options sub-TLV
+    if (aTypeIdFlagsCount > 0)
+    {
+        linkMetricsQueryOptionsTlv.Init();
+        linkMetricsQueryOptionsTlv.SetLength(aTypeIdFlagsCount * sizeof(LinkMetricsTypeIdFlags));
+
+        memcpy(buf + length, &linkMetricsQueryOptionsTlv, sizeof(linkMetricsQueryOptionsTlv));
+        length += sizeof(linkMetricsQueryOptionsTlv);
+        memcpy(buf + length, aTypeIdFlags, linkMetricsQueryOptionsTlv.GetLength());
+        length += linkMetricsQueryOptionsTlv.GetLength();
+    }
+
+    // Set Length for Link Metrics Report Tlv
+    tlv->SetLength(length - sizeof(Tlv));
+
+    SuccessOrExit(error = Get<Mle::MleRouter>().SendDataRequest(aDestination, tlvs, sizeof(tlvs), 0, buf, length));
+
+exit:
+    return error;
+}
+
+otError LinkMetrics::AppendSingleProbeLinkMetricsReport(Message &                     aMessage,
+                                                        uint8_t &                     aLength,
+                                                        const LinkMetricsTypeIdFlags *aTypeIdFlags,
+                                                        uint8_t                       aTypeIdFlagsCount,
+                                                        const int8_t                  aNoiseFloor,
+                                                        const Message &               aRequestMessage)
+{
+    otError                 error = OT_ERROR_NONE;
+    LinkMetricsReportSubTlv metric;
+
+    OT_ASSERT(aTypeIdFlags != nullptr);
+
+    for (uint8_t i = 0; i < aTypeIdFlagsCount; i++)
+    {
+        // Link Metrics Report sub-TLVs
+        metric.Init();
+        metric.SetMetricsTypeId(aTypeIdFlags[i]);
+
+        switch (aTypeIdFlags[i].GetMetricEnum())
+        {
+        case OT_LINK_METRICS_PDU_COUNT:
+            VerifyOrExit(aTypeIdFlags[i].IsLengthFlagSet(), error = OT_ERROR_INVALID_ARGS);
```

Comment:
Here we again verify that the passed-in `aTypeIdFlags` values are valid (i.e. combinations that we support), but we already checked them when we parse the request message.

I suggest instead of passing this as an array of `aTypeIdFlags[]`, we pass a `otLinkMetrics` instance, we can then use the constant `kTypeIdPduCount `, `kTypeIdLqi`, etc (i.e. valid combinations of TypeIdFlags we support) to prepare the sub-TLVs.